dm/relay/local_reader.go

```diff
@@ -585,19 +588,24 @@ func (r *BinlogReader) parseFile(
 		}
 
 		// align with MySQL
-		// if an event's gtid has been contained by given gset
-		// replace it with HEARTBEAT event
-		// for Mariadb, it will bee replaced with MARIADB_GTID_LIST_EVENT
-		// In DM, we replace both of them with HEARTBEAT event
-		if state.replaceWithHeartbeat {
+		// ref https://github.com/pingcap/tiflow/issues/5063#issuecomment-1082678211
+		// heartbeat period is implemented in LocalStreamer.GetEvent
+		if state.skipGTID {
 			switch e.Event.(type) {
 			// Only replace transaction event
 			// Other events such as FormatDescriptionEvent, RotateEvent, etc. should be the same as before
 			case *replication.RowsEvent, *replication.QueryEvent, *replication.GTIDEvent, *replication.XIDEvent, *replication.TableMapEvent:
 				// replace with heartbeat event
-				e = event.GenHeartbeatEvent(e.Header)
+				state.lastSkipGTIDHeader = e.Header
 			default:
 			}
+			return nil
+		} else if lastSkipGTID && state.lastSkipGTIDHeader != nil {
+			// skipGTID is turned off after this event
+			select {
+			case s.ch <- event.GenHeartbeatEvent(state.lastSkipGTIDHeader):
+			case <-ctx.Done():
```

Comment:
prefer remove this branch, since we don't have strong dependency on heartbeat event and `LocalStreamer.GetEvent` has simulate normal heartbeat.

Reply:
mainly to reduce the behaving difference with MySQL, the original purpose is stated in https://github.com/pingcap/dm/pull/1390